<!DOCTYPE html>
<html>

<head>

<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>CP3</title>


<style type="text/css">
body {
  font-family: Helvetica, arial, sans-serif;
  font-size: 14px;
  line-height: 1.6;
  padding-top: 10px;
  padding-bottom: 10px;
  background-color: white;
  padding: 30px; }

body > *:first-child {
  margin-top: 0 !important; }
body > *:last-child {
  margin-bottom: 0 !important; }

a {
  color: #4183C4; }
a.absent {
  color: #cc0000; }
a.anchor {
  display: block;
  padding-left: 30px;
  margin-left: -30px;
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0; }

h1, h2, h3, h4, h5, h6 {
  margin: 20px 0 10px;
  padding: 0;
  font-weight: bold;
  -webkit-font-smoothing: antialiased;
  cursor: text;
  position: relative; }

h1:hover a.anchor, h2:hover a.anchor, h3:hover a.anchor, h4:hover a.anchor, h5:hover a.anchor, h6:hover a.anchor {
  background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA09pVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMy1jMDExIDY2LjE0NTY2MSwgMjAxMi8wMi8wNi0xNDo1NjoyNyAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNiAoMTMuMCAyMDEyMDMwNS5tLjQxNSAyMDEyLzAzLzA1OjIxOjAwOjAwKSAgKE1hY2ludG9zaCkiIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OUM2NjlDQjI4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OUM2NjlDQjM4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo5QzY2OUNCMDg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo5QzY2OUNCMTg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PsQhXeAAAABfSURBVHjaYvz//z8DJYCRUgMYQAbAMBQIAvEqkBQWXI6sHqwHiwG70TTBxGaiWwjCTGgOUgJiF1J8wMRAIUA34B4Q76HUBelAfJYSA0CuMIEaRP8wGIkGMA54bgQIMACAmkXJi0hKJQAAAABJRU5ErkJggg==) no-repeat 10px center;
  text-decoration: none; }

h1 tt, h1 code {
  font-size: inherit; }

h2 tt, h2 code {
  font-size: inherit; }

h3 tt, h3 code {
  font-size: inherit; }

h4 tt, h4 code {
  font-size: inherit; }

h5 tt, h5 code {
  font-size: inherit; }

h6 tt, h6 code {
  font-size: inherit; }

h1 {
  font-size: 28px;
  color: black; }

h2 {
  font-size: 24px;
  border-bottom: 1px solid #cccccc;
  color: black; }

h3 {
  font-size: 18px; }

h4 {
  font-size: 16px; }

h5 {
  font-size: 14px; }

h6 {
  color: #777777;
  font-size: 14px; }

p, blockquote, ul, ol, dl, li, table, pre {
  margin: 15px 0; }

hr {
  background: transparent url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAYAAAAECAYAAACtBE5DAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAyJpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNSBNYWNpbnRvc2giIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OENDRjNBN0E2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OENDRjNBN0I2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo4Q0NGM0E3ODY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo4Q0NGM0E3OTY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PqqezsUAAAAfSURBVHjaYmRABcYwBiM2QSA4y4hNEKYDQxAEAAIMAHNGAzhkPOlYAAAAAElFTkSuQmCC) repeat-x 0 0;
  border: 0 none;
  color: #cccccc;
  height: 4px;
  padding: 0;
}

body > h2:first-child {
  margin-top: 0;
  padding-top: 0; }
body > h1:first-child {
  margin-top: 0;
  padding-top: 0; }
  body > h1:first-child + h2 {
    margin-top: 0;
    padding-top: 0; }
body > h3:first-child, body > h4:first-child, body > h5:first-child, body > h6:first-child {
  margin-top: 0;
  padding-top: 0; }

a:first-child h1, a:first-child h2, a:first-child h3, a:first-child h4, a:first-child h5, a:first-child h6 {
  margin-top: 0;
  padding-top: 0; }

h1 p, h2 p, h3 p, h4 p, h5 p, h6 p {
  margin-top: 0; }

li p.first {
  display: inline-block; }
li {
  margin: 0; }
ul, ol {
  padding-left: 30px; }

ul :first-child, ol :first-child {
  margin-top: 0; }

dl {
  padding: 0; }
  dl dt {
    font-size: 14px;
    font-weight: bold;
    font-style: italic;
    padding: 0;
    margin: 15px 0 5px; }
    dl dt:first-child {
      padding: 0; }
    dl dt > :first-child {
      margin-top: 0; }
    dl dt > :last-child {
      margin-bottom: 0; }
  dl dd {
    margin: 0 0 15px;
    padding: 0 15px; }
    dl dd > :first-child {
      margin-top: 0; }
    dl dd > :last-child {
      margin-bottom: 0; }

blockquote {
  border-left: 4px solid #dddddd;
  padding: 0 15px;
  color: #777777; }
  blockquote > :first-child {
    margin-top: 0; }
  blockquote > :last-child {
    margin-bottom: 0; }

table {
  padding: 0;border-collapse: collapse; }
  table tr {
    border-top: 1px solid #cccccc;
    background-color: white;
    margin: 0;
    padding: 0; }
    table tr:nth-child(2n) {
      background-color: #f8f8f8; }
    table tr th {
      font-weight: bold;
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr td {
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr th :first-child, table tr td :first-child {
      margin-top: 0; }
    table tr th :last-child, table tr td :last-child {
      margin-bottom: 0; }

img {
  max-width: 100%; }

span.frame {
  display: block;
  overflow: hidden; }
  span.frame > span {
    border: 1px solid #dddddd;
    display: block;
    float: left;
    overflow: hidden;
    margin: 13px 0 0;
    padding: 7px;
    width: auto; }
  span.frame span img {
    display: block;
    float: left; }
  span.frame span span {
    clear: both;
    color: #333333;
    display: block;
    padding: 5px 0 0; }
span.align-center {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-center > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: center; }
  span.align-center span img {
    margin: 0 auto;
    text-align: center; }
span.align-right {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-right > span {
    display: block;
    overflow: hidden;
    margin: 13px 0 0;
    text-align: right; }
  span.align-right span img {
    margin: 0;
    text-align: right; }
span.float-left {
  display: block;
  margin-right: 13px;
  overflow: hidden;
  float: left; }
  span.float-left span {
    margin: 13px 0 0; }
span.float-right {
  display: block;
  margin-left: 13px;
  overflow: hidden;
  float: right; }
  span.float-right > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: right; }

code, tt {
  margin: 0 2px;
  padding: 0 5px;
  white-space: nowrap;
  border: 1px solid #eaeaea;
  background-color: #f8f8f8;
  border-radius: 3px; }

pre code {
  margin: 0;
  padding: 0;
  white-space: pre;
  border: none;
  background: transparent; }

.highlight pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }

pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }
  pre code, pre tt {
    background-color: transparent;
    border: none; }

sup {
    font-size: 0.83em;
    vertical-align: super;
    line-height: 0;
}

kbd {
  display: inline-block;
  padding: 3px 5px;
  font-size: 11px;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb
}

* {
	-webkit-print-color-adjust: exact;
}
@media screen and (min-width: 914px) {
    body {
        width: 854px;
        margin:0 auto;
    }
}
@media print {
	table, pre {
		page-break-inside: avoid;
	}
	pre {
		word-wrap: break-word;
	}
  body {
    padding: 2cm; 
  }
}
</style>


</head>

<body>

<h1 id="toc_0"><p align="center">Capstone Project III: Biological Applications of Regular Expressions</p></h1>

<p align="center"> <b> BEFORE STARTING THIS ASSIGNMENT, PLEASE READ ALL OF THE INSTRUCTIONS</b></p>

<h2 id="toc_1">Prerequisites</h2>

<ul>
<li>Prior to starting this project, you should feel comfortable constructing regular expressions to parse text</li>
<li>Completion of the Regex Quiz with at least 80% or higher</li>
<li>Completion of Lab 6 and a solid understanding of the listed learning objectives </li>
</ul>

<h2 id="toc_2">Learning Goals</h2>

<ul>
<li>Get practice using <code>awk</code> and <code>sed</code> with regular expresssions to parse real biological datasets</li>
<li>Get practice completing computing tasks without step-by-step instructions. As mentioned on Capstone Project #2, you will often have a specific scientific goal rather than a set of instructions for how to obtain the results. Here you will get practice building your OWN workflow for a data analysis project.</li>
<li>Reinforcement of the concept of doing projects in a paced setting that involves mutliple sessions working on various pieces and returning after breaks and reflection</li>
<li>As you can see from the rubric for this assignment, you will no longer be graded on your effort, but instead on your output. While you will still turn in your history (or a script), we will not be grading it for commands as was done with Capstone #1. Instead, you will be graded on whether you get the CORRECT answers in your final tables.</li>
</ul>

<h2 id="toc_3">Learning Objectives</h2>

<ul>
<li>Get exposure to genomic datasets of genetic variants</li>
<li>Reinforce the use of awk to count records in datasets, isolate specific columns, and perform calculations</li>
<li>Reinforce printing tabular outputs from STDOUT</li>
<li>Incrementally build one-liners using pipes either in a script or on the CL</li>
<li>Get practice building regular expressions to match patterns in DNA</li>
<li>Get practice using sed to find and replace patterns in files</li>
</ul>

<h2 id="toc_4">Step-by-Step Instructions</h2>

<h6 id="toc_5">Instructions: Often when dealing with biological problems, you are not given step-by-step instructions as you have gotten used to throughout this course. Instead, you have a specific question that you need to answer and you need to decide which approach is most useful to get the data you need. For your 3rd Capstone Project, I will not tell you the tools you need to accomplish the goals, only what the problems are that you need to solve. As with other assignments, you will need to &quot;show your work&quot; in the form of clearing your history at the beginning of the project and saving it to a file when you are done. If you work on this intermittenly, please remove non-relevant lines from your history before turning it in for a grade (we will be grading presence/absence, not content). While you may work in whatever linux environment you wish, if you plan to use sed, please note that it often has unexpected behavior on MacOS. See the Week 9 canvas page for how to get sed to behave on Mac OS.</h6>

<h5 id="toc_6">Assignment: This project centers around the use of regular expressions, which we have focused on throughout this module. To complete this assignment, you may choose to write a shell scripts or use a series of one-liners in awk/sed. You may also choose to use a combination of these options to facilitate your goals. Nonetheless, start by making a directory called &quot;YOUR_LAST_NAME_Project3&quot;.</h5>

<h3 id="toc_7">Part A:</h3>

<ol>
<li><p>The file you will be working with for Part A is an output file from a program called &quot;<a href="http://breakdancer.sourceforge.net/">Breakdancer</a>&quot;. For Part A, make a directory called &quot;BD&quot; within your Project3 directory. Copy the <a href="https://auburn.instructure.com/files/224786655/download?download_frd=1">BD output</a> file from canvas to this new folder. As with similar files, header lines start with a <code>#</code>. The most important part of the header is this line:</p>

<div><pre><code class="language-none">#Chr1   Pos1    Orientation1    Chr2    Pos2    Orientation2    Type    Size    Score   num_Reads       num_Reads_lib   M_Arctoides.chr6.recalibrated.bam</code></pre></div></li>
<li><p>Breakdancer takes a genome alignment file (<a href="http://genome.sph.umich.edu/wiki/BAM">bam</a> format) as input and outputs potential <a href="https://www.ncbi.nlm.nih.gov/dbvar/content/overview/">structural variants</a>. We will sift through the raw output to identify high quality variants. Specifically, breakdancer identifies five possible structural variants (SVs): Insertions (INS), Deletions (DEL), Inversions (INV), and Translocations (ITX/CTX). The first thing you need to do is get a count of how many of each <strong><em>type</em></strong> of variant is in this file. </p>

<table>
<thead>
<tr>
<th>Variant Type</th>
<th>Count in File</th>
<th>Filtered count (see #4)</th>
</tr>
</thead>

<tbody>
<tr>
<td>Insertions</td>
<td></td>
<td></td>
</tr>
<tr>
<td>Deletions</td>
<td></td>
<td></td>
</tr>
<tr>
<td>Inversions</td>
<td></td>
<td></td>
</tr>
<tr>
<td>Translocations (ITX)</td>
<td></td>
<td></td>
</tr>
<tr>
<td>Translocations (CTX)</td>
<td></td>
<td></td>
</tr>
</tbody>
</table></li>
<li><p>The next thing you need to do is to gain a better understanding of each type of variant. To accomplish this, calculate the mean value for each of the following columns <em>per variant type</em>: Size, Score, num_Reads. </p>

<p>Special note for <strong>Size</strong>: The column for size has both negative and positive numbers, making it difficult to calculate a mean value. Also, several of these values are just plain WRONG! Therefore to get the mean, you will need to <strong>re-calculate</strong> the length of each variant from the Pos1 and Pos2 columns. Depending on the tool you use to do this calculation, you may have issues with it printing, for which you are allowed to use the internet. <em>For translocations between chromosomes, you can ignore the fact that the coordinates are on different chromosomes and do the calculations the same as other types of variants.</em></p>

<table>
<thead>
<tr>
<th>Variant Type</th>
<th>Mean Score</th>
<th>Mean Size</th>
<th>Mean # Reads</th>
</tr>
</thead>

<tbody>
<tr>
<td>Insertions</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>Deletions</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>Inversions</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>Translocations (ITX)</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>Translocations (CTX)</td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table></li>
<li><p>Finally, we will want to filter for high quality variants. Let&#39;s repeat the Table we made in #2, but only for filtered variants. First, <strong>remove</strong> all variants with a score below 95. The next filter we will want to impose is based on the number of reads. In addition to the score filter, now require that our trusted variants are supported by at least 10 reads.  Go back and fill in the count for each type based on these two filters combined. </p></li>
</ol>

<h3 id="toc_8">Part B:</h3>

<ol>
<li><p>For part B of your project, you will be acting as a restriction enzyme. <a href="https://en.wikipedia.org/wiki/Restriction_enzyme">Restriction enzymes</a> cleave DNA based on a recognition sequence. Retrieve the fasta sequence from CANVAS. Make a new directory in your Project3 folder called &quot;Enzyme&quot; and download this zipped <a href="https://auburn.instructure.com/files/224786594/download?download_frd=1">fasta file</a> into it. Unzip the file.</p></li>
<li><p>Your goal will be to read in the sequence and &quot;cut&quot; the DNA (i.e. insert a newline) everytime you encounter a given cleavage site. Because the purpose of this assignment is to practice your regular expression usage, assume all enzymes cleave with blunt ends. Don&#39;t get fancy here folks! You should think carefully about which tools you&#39;ve learned in class are appropriate for this exercise. It may be easier to first identify the number of cleavage sites within the DNA sample prior to cutting. You may find that more than one tool will work, so go with whichever one you feel most confident using. To avoid getting the number of cut sites wrong, consider removing the fasta header from the sequence file.</p></li>
<li><p>One of the most common restriction enzymes is EcoRI, which has a recognition site of <code>GAATTC</code>. Fill in the table below on the number of resulting fragments from EcoR1 restriction enzyme.  </p>

<table>
<thead>
<tr>
<th>Restriction Enzyme</th>
<th>Cleaveage Site</th>
<th>Number of Fragments</th>
<th>Average Fragment Length</th>
</tr>
</thead>

<tbody>
<tr>
<td>EcoR1</td>
<td><code>GAATTC</code></td>
<td></td>
<td></td>
</tr>
</tbody>
</table></li>
<li><p>Now that you have cleaved the larger DNA sequence file into multiple lines, compute the average length of the resulting fragments you made <strong><em>&quot;in silico&quot;</em></strong>. <em>You may search the internet for options to do this</em>. Add the result to the table in #3 above. </p>

<blockquote>
<p>Note: the key will allow all values to be off by +/- 1. No partial credit will be awarded for table values - they are either right or wrong. So consider making test files where you KNOW the answer to test that your code works as expected. </p>
</blockquote></li>
<li><p>Two other restriction enzymes have more variable restriction sites. In DNA, <a href="http://www.bioinformatics.org/sms/iupac.html">ambiguous bases</a> are indicated via specific codes. As you did above, repeat the process with these two new restriction sites. Unlike EcoRI, the variability will require you to generate a regular expression based on their site ambiguity in order to cleave them. As above, once you cleave them, be sure to calculate the average fragment length as well.</p>

<table>
<thead>
<tr>
<th>Restriction Enzyme</th>
<th>Cleavage Site</th>
<th>Number of Fragments</th>
<th>Average Fragment Length</th>
</tr>
</thead>

<tbody>
<tr>
<td>AcyI</td>
<td><code>GRCGYC</code></td>
<td></td>
<td></td>
</tr>
<tr>
<td>AhdI</td>
<td><code>GACNNNNNGTC</code></td>
<td></td>
<td></td>
</tr>
</tbody>
</table></li>
</ol>

<h3 id="toc_9">Part C:</h3>

<ol>
<li><p>Congrats on making it through this project! Now that you are done, convert your tables into text files as you did in Lab #6. </p></li>
<li><p>You should have two tables for Part A (each value is worth 1 pt, totalling 25 across both tables). Combine the EcoRI results with the other restriction enzymes to create one table for Part B (each value is worth 2 pts, totalling 12 points for Table 3). Put the three files in the main project directory, not the subdirectories.</p></li>
<li><p>Save your history/script to the main directory and move up a directory to make a tarball of this directory. Upload your tarball to Canvas. </p></li>
</ol>

<h2 id="toc_10">Reflection Questions (4 pts)</h2>

<h5 id="toc_11">Lastly, respond to the following reflection questions, and upload your responses titled <code>NAME_reflection.txt</code> to canvas along with your tarball. In your reflection, answer the following questions:</h5>

<ol>
<li><p>Many programs written for biological data are for parsing and finding different signatures in the data. Describe some biological signatures, or <strong>patterns</strong>, common to your field of biology and how they are computationally detected (likely through regex) and represented (i.e., file types &amp; data structures).</p></li>
<li><p>Now having seen several file formats in this course (e.g. museum records/fasta/VCF/etc), what are some features of various formats you have encountered that lend themselves to different types of data analysis you have done in the course?</p>

<h4 id="toc_12">This capstone project completes your basic training on the command line. (We will return to it briefly at the end of Module 4 and for the final!) The next two questions ask you to reflect on these first 10 weeks of the course:</h4></li>
<li><p>In the first 3 modules of the course, what topics were <strong>not</strong> covered that you wish you had learned? It is OK to say you feel like you have learned ENOUGH!</p></li>
<li><p>In the beginning of the semester, I made the analogy that this course is like learning a foreign language. Many study abroad students find they lose their language skills when they return home unless they dedicate time to keeping their skills sharp. <strong>What are some things YOU could do to continue your training on command line so as not to lose the skills you have learned in the last 10 weeks?</strong></p></li>
</ol>




</body>

</html>
